File: logistics/logistics/report/cargo_log_report/cargo_log_report.js
// Copyright (c) 2024, siva and contributors
// For license information, please see license.txt

frappe.query_reports["Cargo Log Report"] = {
    "filters": [
        {
            "fieldname": "mode_of_transport",
            "label": __("Mode of Transport"),
            "fieldtype": "Select",
            "options": "\nAir\nSea",
            "default": ""  
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_days(frappe.datetime.nowdate(), -30)  
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.nowdate() 
        },
    ],

    "onload": function(report) {
        report.page.add_inner_button(__("Refresh"), function() {
            report.refresh();
        });
    },

    "formatter": function(value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname === "total_profit") {
            if (data.total_profit < 0) {
                value = `<span style="color:red;">${value}</span>`;
            } else {
                value = `<span style="color:green;">${value}</span>`;
            }
        }

        return value;
    },

    "get_data": function(filters) {
        return frappe.call({
            method: "frappe.desk.query_report.get_data",
            args: {
                report_name: "Cargo Log Report",
                filters: filters
            }
        }).then(function(response) {
            return response.message;
        });
    }
};

File: logistics/logistics/report/cargo_log_report/cargo_log_report.py
# Copyright (c) 2024, siva and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def get_conditions(filters):
    conditions = {}

    if filters.get('from_date') and filters.get('to_date'):
        conditions['date'] = ['between', [filters.get('from_date'), filters.get('to_date')]]
    elif filters.get('from_date'):
        conditions['date'] = ['>=', filters.get('from_date')]
    elif filters.get('to_date'):
        conditions['date'] = ['<=', filters.get('to_date')]

    if filters.get('mode_of_transport'):
        conditions['mode_of_transport'] = filters.get('mode_of_transport')

    return conditions

def execute(filters=None):
    columns = [
        {"label": "Mode of Transport", "fieldname": "mode_of_transport", "fieldtype": "Data", "width": 200},
        {"label": "Total Cost", "fieldname": "total_cost", "fieldtype": "Currency", "width": 150},
        {"label": "Total Profit", "fieldname": "total_profit", "fieldtype": "Currency", "width": 150},
        {"label": "Total Charges", "fieldname": "total_charge", "fieldtype": "Currency", "width": 150},
    ]

    conditions = get_conditions(filters)

    data = frappe.get_all(
        'Cargo Log',
        fields=[
            'mode_of_transport',
            'SUM(cost_price) as total_cost',
            'SUM(profit) as total_profit',
            'SUM(charges) as total_charge'
        ],
        filters=conditions,
        group_by='mode_of_transport',
        # as_dict=True
    )

    if not data:
        frappe.msgprint(_('No records found'))

    return columns, data
